From 258a7fa9b3b80479c43ab16d24ea3f41c966e421 Mon Sep 17 00:00:00 2001
From: Dinesh Sudham <dineshx.sudham@intel.com>
Date: Tue, 27 Oct 2020 13:48:18 +0800
Subject: [PATCH] PONRTSYS:GSWAPI: Delete bridge port egress meters when its
 disabled

Meter index should be released when its disabled or not assigned in
bridge port.

Signed-off-by: Dinesh Sudham <dineshx.sudham@intel.com>
---
 .../net/ethernet/lantiq/switch-api/gsw_flow_core.c | 63 ++++++++++++++++++++++
 1 file changed, 63 insertions(+)

diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
index ad60bdc8afc6..fc3a34a0d2d3 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
@@ -23194,6 +23194,17 @@ GSW_return_t GSW_BridgePortConfigSet(void *cdev, GSW_BRIDGE_portConfig_t *param)
 		} else {
 			/*Disable Broadcast sub meter*/
 			tbl_prog_brdgeport_egress.val[5] &= ~(1 << 7);
+			if (brdge_pcfg->BroadcastMeteringAssigned) {
+				idx = brdge_pcfg->BroadcastMeteringId;
+				if (idx >= ARRAY_SIZE(gswdev->meter_idx)) {
+					pr_err("ERROR :idx for meter_idx %d out of bounds!", idx);
+					ret = GSW_statusErr;
+					goto UNLOCK_AND_RETURN;
+				}
+				gswdev->meter_idx[idx].IndexInUsageCnt--;
+				brdge_pcfg->BroadcastMeteringAssigned = 0;
+				brdge_pcfg->BroadcastMeteringId = 0;
+			}
 		}
 
 		/*GSW_BRIDGE_PORT_EGRESS_METER_MULTICAST*/
@@ -23255,6 +23266,17 @@ GSW_return_t GSW_BridgePortConfigSet(void *cdev, GSW_BRIDGE_portConfig_t *param)
 		} else {
 			/*Disable Multicast sub meter*/
 			tbl_prog_brdgeport_egress.val[6] &= ~(1 << 7);
+			if (brdge_pcfg->MulticastMeteringAssigned) {
+				idx = brdge_pcfg->MulticastMeteringId;
+				if (idx >= ARRAY_SIZE(gswdev->meter_idx)) {
+					pr_err("ERROR :idx for meter_idx %d out of bounds!", idx);
+					ret = GSW_statusErr;
+					goto UNLOCK_AND_RETURN;
+				}
+				gswdev->meter_idx[idx].IndexInUsageCnt--;
+				brdge_pcfg->MulticastMeteringAssigned = 0;
+				brdge_pcfg->MulticastMeteringId = 0;
+			}
 		}
 
 		/*GSW_BRIDGE_PORT_EGRESS_METER_UNKNOWN_MC_IP*/
@@ -23317,6 +23339,17 @@ GSW_return_t GSW_BridgePortConfigSet(void *cdev, GSW_BRIDGE_portConfig_t *param)
 		} else {
 			/*Disable Unknown Multicast ip sub meter*/
 			tbl_prog_brdgeport_egress.val[9] &= ~(1 << 7);
+			if (brdge_pcfg->UnknownMultiIpMeteringAssigned) {
+				idx = brdge_pcfg->UnknownMultiIpMeteringId;
+				if (idx >= ARRAY_SIZE(gswdev->meter_idx)) {
+					pr_err("ERROR :idx for meter_idx %d out of bounds!", idx);
+					ret = GSW_statusErr;
+					goto UNLOCK_AND_RETURN;
+				}
+				gswdev->meter_idx[idx].IndexInUsageCnt--;
+				brdge_pcfg->UnknownMultiIpMeteringAssigned = 0;
+				brdge_pcfg->UnknownMultiIpMeteringId = 0;
+			}
 		}
 
 		/*GSW_BRIDGE_PORT_EGRESS_METER_UNKNOWN_MC_NON_IP*/
@@ -23378,6 +23411,17 @@ GSW_return_t GSW_BridgePortConfigSet(void *cdev, GSW_BRIDGE_portConfig_t *param)
 		} else {
 			/*Disable Unknown Multicast Non-ip sub meter*/
 			tbl_prog_brdgeport_egress.val[8] &= ~(1 << 7);
+			if (brdge_pcfg->UnknownMultiNonIpMeteringAssigned) {
+				idx = brdge_pcfg->UnknownMultiNonIpMeteringId;
+				if (idx >= ARRAY_SIZE(gswdev->meter_idx)) {
+					pr_err("ERROR :idx for meter_idx %d out of bounds!", idx);
+					ret = GSW_statusErr;
+					goto UNLOCK_AND_RETURN;
+				}
+				gswdev->meter_idx[idx].IndexInUsageCnt--;
+				brdge_pcfg->UnknownMultiNonIpMeteringAssigned = 0;
+				brdge_pcfg->UnknownMultiNonIpMeteringId = 0;
+			}
 		}
 
 		/*GSW_BRIDGE_PORT_EGRESS_METER_UNKNOWN_UC*/
@@ -23439,6 +23483,17 @@ GSW_return_t GSW_BridgePortConfigSet(void *cdev, GSW_BRIDGE_portConfig_t *param)
 		} else {
 			/*Disable Unknown Unicast sub meter*/
 			tbl_prog_brdgeport_egress.val[7] &= ~(1 << 7);
+			if (brdge_pcfg->UnknownUniCastMeteringAssigned) {
+				idx = brdge_pcfg->UnknownUniCastMeteringId;
+				if (idx >= ARRAY_SIZE(gswdev->meter_idx)) {
+					pr_err("ERROR :idx for meter_idx %d out of bounds!", idx);
+					ret = GSW_statusErr;
+					goto UNLOCK_AND_RETURN;
+				}
+				gswdev->meter_idx[idx].IndexInUsageCnt--;
+				brdge_pcfg->UnknownUniCastMeteringAssigned = 0;
+				brdge_pcfg->UnknownUniCastMeteringId = 0;
+			}
 		}
 
 		/*GSW_BRIDGE_PORT_EGRESS_METER_OTHERS*/
@@ -26808,6 +26863,14 @@ GSW_return_t GSW_QOS_MeterFree(void *cdev, GSW_QoS_meterCfg_t *param)
 
 	/*Just mark the Meter idx as free*/
 	gswdev->meter_idx[param->nMeterId].IndexInUse = 0;
+	/* Meter Instance Selection*/
+	gsw_w32(cdev, GSW_INST_SEL_INST_OFFSET,
+		GSW_INST_SEL_INST_SHIFT,
+		GSW_INST_SEL_INST_SIZE, param->nMeterId);
+	/* Disable the meter shaper */
+	gsw_w32(cdev, GSW_PCE_TCM_CTRL_TCMEN_OFFSET,
+		GSW_PCE_TCM_CTRL_TCMEN_SHIFT,
+		GSW_PCE_TCM_CTRL_TCMEN_SIZE, 0);
 	ret = GSW_statusOk;
 
 UNLOCK_AND_RETURN:
